repos/system_upgrade/common/actors/checkdetecteddevicesanddrivers/libraries/checkdddd.py

```diff
@@ -0,0 +1,180 @@
+from leapp.models import DetectedDeviceOrDriver
+
+
+class MessagingClass(IntEnum):
+    UNKNOWN = 0
+    DRIVERS = 1
+    DEVICES = 2
+    CPUS = 3
+
+
+def create_inhibitors(inhibiting_entries):
+    if not inhibiting_entries:
+        return
+
+    drivers = inhibiting_entries.get(MessagingClass.DRIVERS)
+    if drivers:
+        reporting.create_report([
+            reporting.Title(
+                'Leapp detected loaded kernel drivers which have been removed '
+                'in RHEL {}. Upgrade cannot proceed.'.format(get_target_major_version())
+            ),
+            reporting.Summary(
+                (
+                    'Support for the following RHEL {source} device drivers has been removed in RHEL {target}:\n'
+                    '     - {drivers}\n'
+                ).format(
+                    drivers='\n     - '.join([entry.driver_name for entry in drivers]),
+                    target=get_target_major_version(),
+                    source=get_source_major_version(),
+                )
+            ),
+            reporting.Audience('sysadmin'),
+            reporting.Tags([reporting.Tags.KERNEL, reporting.Tags.DRIVERS]),
+            reporting.Severity(reporting.Severity.HIGH),
+            reporting.Flags([reporting.Flags.INHIBITOR])
+        ])
+
+    devices = inhibiting_entries.get(MessagingClass.DEVICES)
+    if devices:
+        reporting.create_report([
+            reporting.Title(
+                'Leapp detected devices which are no longer supported in RHEL {}. Upgrade cannot proceed.'.format(
+                    get_target_major_version())
+            ),
+            reporting.Summary(
+                (
+                    'Support for the following devices has been removed in RHEL {target}:\n'
+                    '     - {devices}\n'
+                ).format(
+                    devices='\n     - '.join(['{name} ({pci})'.format(name=entry.device_name,
+                                             pci=entry.device_id) for entry in devices]),
+                    target=get_target_major_version(),
+                )
+            ),
+            reporting.Audience('sysadmin'),
+            reporting.Tags([reporting.Tags.KERNEL]),
+            reporting.Severity(reporting.Severity.HIGH),
+            reporting.Flags([reporting.Flags.INHIBITOR])
+        ])
+
+    cpu = inhibiting_entries.get(MessagingClass.CPUS)
+    if cpu:
+        reporting.create_report([
+            reporting.Title(
+                'Leapp detected a processor which is no longer supported in RHEL {}. Upgrade cannot proceed.'.format(
+                    get_target_major_version())
+            ),
+            reporting.Summary(
+                (
+                    'Support for the following processors has been removed in RHEL {target}:\n'
+                    '     - {processors}\n'
+                ).format(
+                    processors='\n     - '.join([entry.device_name for entry in devices]),
```

Comment:
This causes the traceback because devices is None
```
Traceback (most recent call last):
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib64/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python3.6/site-packages/leapp/repository/actor_definition.py", line 72, in _do_run
    actor_instance.run(*args, **kwargs)
  File "/usr/lib/python3.6/site-packages/leapp/actors/__init__.py", line 290, in run
    self.process(*args)
  File "/etc/leapp/repos.d/system_upgrade/common/actors/checkdetecteddevicesanddrivers/actor.py", line 21, in process
    checkdddd.process()
  File "/etc/leapp/repos.d/system_upgrade/common/actors/checkdetecteddevicesanddrivers/libraries/checkdddd.py", line 179, in process
    create_inhibitors(inhibiting)
  File "/etc/leapp/repos.d/system_upgrade/common/actors/checkdetecteddevicesanddrivers/libraries/checkdddd.py", line 79, in create_inhibitors
    processors='\n     - '.join([entry.device_name for entry in devices]),
TypeError: 'NoneType' object is not iterable
```

leapp-upgrade-el8toel9-0.15.0-0.202203071217Z.1929ed9.HEAD.PR835.el8.noarch

Reply:
`s/devices/cpu` got to me to 
```
============================================================
                     UPGRADE INHIBITED                      
============================================================

Upgrade has been inhibited due to the following problems:
    1. Inhibitor: Leapp detected a processor which is no longer supported in RHEL 9. Upgrade cannot proceed.
Consult the pre-upgrade report for details and possible remediation.
```

```
Risk Factor: high (inhibitor)
Title: Leapp detected a processor which is no longer supported in RHEL 9. Upgrade cannot proceed.
Summary: Support for the following processors has been removed in RHEL 9:
     - SKYLAKE_X

Key: f79672290945c09de12a14b28906b98d5c8ed68a
```

Reply:
Well according to the deprecation data it has no support on rhel9 anymore then

Reply:
damn, unsaved changes which haven't been committed, sorry about that.